packages/react-reconciler/src/ReactFiberCommitWork.js

```diff
@@ -295,8 +295,21 @@ function commitHookEffectList(
       if ((effect.tag & mountTag) !== NoHookEffect) {
         // Mount
         const create = effect.create;
-        const destroy = create();
-        effect.destroy = typeof destroy === 'function' ? destroy : null;
+        let destroy = create();
+        if (typeof destroy !== 'function') {
+          if (__DEV__ && destroy !== null && destroy !== undefined) {
```

Comment:
Nit: I kinda perfer `if (__DEV__)` to be a separate line for easier visual scannning